daemon/actual_ip_resolver.go

```diff
@@ -0,0 +1,131 @@
+package daemon
+
+import (
+	"context"
+	"log"
+	"net/netip"
+	"time"
+
+	"github.com/NordSecurity/nordvpn-linux/core"
+	"github.com/NordSecurity/nordvpn-linux/daemon/state"
+	"github.com/NordSecurity/nordvpn-linux/events"
+	"github.com/NordSecurity/nordvpn-linux/internal"
+	"github.com/NordSecurity/nordvpn-linux/network"
+)
+
+// tryInsightsIP is an attempt to get insights IP
+// the IP it returns is always valid
+func tryInsightsIP(api core.InsightsAPI) (netip.Addr, error) {
+	insights, err := api.InsightsViaTunnel()
+	if err != nil || insights == nil {
+		return netip.Addr{}, err
+	}
+
+	ip, err := netip.ParseAddr(insights.IP)
+	if err == nil {
+		return ip, nil
+	} else {
+		return netip.Addr{}, err
+	}
+}
+
+func insightsIPUntilSuccess(ctx context.Context, api core.InsightsAPI, backoff func(int) time.Duration) (netip.Addr, error) {
+	type Result struct {
+		netip.Addr
+		error
+	}
+	result := make(chan Result, 1)
+	for i := 0; ; i++ {
+		if ctx.Err() != nil {
+			return netip.Addr{}, ctx.Err()
+		}
+
+		// this goroutine is used so that this function is immediately stopped once the context is cancelled
+		go func() {
```

Comment:
Do we need to have this in a separate goroutine? :thinking: What if the `tryInsightsIP` is slow? Won't this cause lot of goroutines to be spawned?

Reply:
no, it won't call `tryInsightsIP` unless the previous call returned.